File: backends/dotnet/services/Geo/Geo.App/Implementations/CQRS/Handlers/Q/CheckHealth.cs
// -----------------------------------------------------------------------
// <copyright file="CheckHealth.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Geo.App.Implementations.CQRS.Handlers.Q;

using D2.Shared.Handler;
using D2.Shared.Result;

using ComponentHealth = D2.Geo.App.Interfaces.CQRS.Handlers.Q.IQueries.ComponentHealth;
using H = D2.Geo.App.Interfaces.CQRS.Handlers.Q.IQueries.ICheckHealthHandler;
using I = D2.Geo.App.Interfaces.CQRS.Handlers.Q.IQueries.CheckHealthInput;
using O = D2.Geo.App.Interfaces.CQRS.Handlers.Q.IQueries.CheckHealthOutput;
using ReadCache = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead;
using ReadRepo = D2.Geo.App.Interfaces.Repository.Handlers.R.IRead;

/// <summary>
/// Handler for checking the overall health of the Geo service by aggregating
/// the results of individual component ping handlers.
/// </summary>
public class CheckHealth : BaseHandler<H, I, O>, H
{
    private readonly ReadRepo.IPingDbHandler r_pingDb;
    private readonly ReadCache.IPingHandler r_pingCache;

    /// <summary>
    /// Initializes a new instance of the <see cref="CheckHealth"/> class.
    /// </summary>
    ///
    /// <param name="pingDb">
    /// The database ping handler.
    /// </param>
    /// <param name="pingCache">
    /// The distributed cache ping handler.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public CheckHealth(
        ReadRepo.IPingDbHandler pingDb,
        ReadCache.IPingHandler pingCache,
        IHandlerContext context)
        : base(context)
    {
        r_pingDb = pingDb;
        r_pingCache = pingCache;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        var components = new Dictionary<string, ComponentHealth>();

        // Run DB and cache pings in parallel.
        // Convert to Task once — ValueTask must not be awaited more than once.
        var dbTask = r_pingDb.HandleAsync(new ReadRepo.PingDbInput(), ct).AsTask();
        var cacheTask = r_pingCache.HandleAsync(new ReadCache.PingInput(), ct).AsTask();

        await Task.WhenAll(dbTask, cacheTask);

        var dbResult = await dbTask;
        if (dbResult.Data is not null)
        {
            components["db"] = new ComponentHealth(
                dbResult.Data.Healthy ? "healthy" : "unhealthy",
                dbResult.Data.LatencyMs,
                dbResult.Data.Error);
        }
        else
        {
            components["db"] = new ComponentHealth("unhealthy", null, "Ping handler returned no data");
        }

        var cacheResult = await cacheTask;
        if (cacheResult.Data is not null)
        {
            components["cache"] = new ComponentHealth(
                cacheResult.Data.Healthy ? "healthy" : "unhealthy",
                cacheResult.Data.LatencyMs,
                cacheResult.Data.Error);
        }
        else
        {
            components["cache"] = new ComponentHealth("unhealthy", null, "Ping handler returned no data");
        }

        var allHealthy = components.Values.All(c => c.Status == "healthy");
        var status = allHealthy ? "healthy" : "degraded";

        return D2Result<O?>.Ok(new O(status, components));
    }
}

File: backends/node/services/files/app/src/implementations/cqrs/handlers/q/check-health.ts
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { Queries } from "../../../../interfaces/cqrs/handlers/index.js";
import { D2Result } from "@d2/result";
import type { IPingDbHandler } from "../../../../interfaces/repository/handlers/index.js";
import type { FileStorageHandlers } from "../../../../interfaces/providers/storage/handlers/index.js";

type Input = Queries.CheckHealthInput;
type Output = Queries.CheckHealthOutput;

/**
 * Aggregates ping results from all files service dependencies into a single
 * health check response.
 */
export class CheckHealth extends BaseHandler<Input, Output> implements Queries.ICheckHealthHandler {
  private readonly pingDb: IPingDbHandler;
  private readonly storage: FileStorageHandlers;

  constructor(pingDb: IPingDbHandler, storage: FileStorageHandlers, context: IHandlerContext) {
    super(context);
    this.pingDb = pingDb;
    this.storage = storage;
  }

  protected async executeAsync(_input: Input): Promise<D2Result<Output | undefined>> {
    const components: Record<string, Queries.ComponentHealth> = {};

    const [dbResult, storageResult] = await Promise.all([
      this.pingDb.handleAsync({}),
      this.storage.ping.handleAsync({}),
    ]);

    // DB
    if (dbResult.data) {
      components["db"] = {
        status: dbResult.data.healthy ? "healthy" : "unhealthy",
        latencyMs: dbResult.data.latencyMs,
        error: dbResult.data.error,
      };
    } else {
      components["db"] = { status: "unhealthy", error: "Ping handler returned no data" };
    }

    // Object storage (MinIO)
    if (storageResult.data) {
      components["storage"] = {
        status: storageResult.data.healthy ? "healthy" : "unhealthy",
        latencyMs: storageResult.data.latencyMs,
        error: storageResult.data.error,
      };
    } else {
      components["storage"] = { status: "unhealthy", error: "Ping handler returned no data" };
    }

    const allHealthy = Object.values(components).every((c) => c.status === "healthy");
    const status = allHealthy ? "healthy" : "degraded";

    return D2Result.ok({ data: { status, components } });
  }
}
